### ISSUE TITLE
[1.4, java api] ExistsRequestBuilder fails to build ExistsRequest

### ISSUE THREAD

**Issue by fredfp** [elasticsearch/elasticsearch#9730]
Using elasticsearch java api 1.4.3, I think I'm experiencing a problem similar to what I reported at https://github.com/yakaz/elasticsearch-action-updatebyquery/issues/27 but this time it is the official API that seem to have the issue:

```scala
val qb = QueryBuilders.constantScoreQuery(FilterBuilders.termFilter("user", "kimchy"))
// returns false, which is expected because I have no such document in my index
client.prepareExists("myIndex")
      .setTypes("myType")
      .setQuery(qb)
      .execute().actionGet().exists()

// returns true... for any query with an empty (but existing) index/type, which is obviously wrong
client.exists(client.prepareExists("myIndex")
      .setTypes("myType")
      .setQuery(qb)
      .request()).actionGet().exists()
```

I think `ExistsRequestBuilder.request()` is the problem here and I need to get an ExistsRequest as it is what my elasticsearch wrapper expects.

**Comment by clintongormley** [comment]
Hi @fredfp 

Have a look at how it is used in the tests: https://github.com/elasticsearch/elasticsearch/blob/master/src/test/java/org/elasticsearch/exists/SimpleExistsTests.java#L62

**Comment by fredfp** [comment]
Thanks @clintongormley

What I'm looking for is a way to get an ExistsRequest correctly built. I need that because I use a wrapper that takes care of asynchronously executing elasticsearch requests.

If I'm not mistaken, the tests you referred to only illustrate the use of the synchronous api and go straight from an ExistsRequestBuilder to a Response. I need an ExistsRequest.

For comparison (and this is why I think it is a bug in this case) I use a similar way to build all my search requests, and it works perfectly fine:
```scala
client.prepareSearch("myIndex")
      .setTypes("myType")
      .setQuery(QueryBuilders.constantScoreQuery(FilterBuilders.termFilter("user", "kimchy")))
      .request()
```

In conclusion, this still looks like a bug to me.

Thoughts?

**Comment by javanna** [comment]
Hi @fredfp I think I see what you mean. The request returned by `request()` in that case might not be complete, as part of it gets set at execution time based on `sourceBuilder` content. Have a look [here](https://github.com/elasticsearch/elasticsearch/blob/5f0719fd508872b57c43226c971a608db8e2856f/src/main/java/org/elasticsearch/action/exists/ExistsRequestBuilder.java#L115). This is a bit annoying, I agree.

May I ask though why you are using builders to retrieve their internal requests and execute them? In general you would use a builder and execute it straight-away instead. In your case, can't you just use requests directly instead of going through the builders step? That seems like a cleaner approach if you want to deal with requests.

**Comment by fredfp** [comment]
Hello @javanna 

As for why I want request objects:
I'm using https://github.com/scalastuff/esclient to execute elasticsearch requests asynchronously and wrap the answer inside a scala Future in a type-safe way. That thing works with Requests in order to know the Response type your are going to get.
I could adapt that lib to make it handle ListenableActionFuture, but I'd prefer to keep that library as the only place where elasticsearch requests are executed.

And why I can't use requests directly:
I would love to only use Requests directly and I already do it for some part of the elasticsearch API, but I found no way to set the query of a search/exists request on a Search/Exists Request object, is there one (equivalent of SearchRequestBuilder.setQuery())?

**Comment by javanna** [comment]
Not sure I get what you mean, `ExistsRequestBuilder` has a couple of `setQuery` methods too. But I also see no difference between `ExistsRequestBuilder` and `SearchRequestBuilder`, they (mis)behave the same way as far as I can see. Can you double check that if you do the same with `SearchRequestBuilder` it works as you'd expect?

As for using builders to get requests back. Whatever you do with a builder you should be able to do with a plain request as well. I'd recommend to stick to requests if that's what you need.

In general, this will all go away with #9201, where we are discussing the path forward for out client api, in order to have a single way to do things and make the api less trappy. I invite you to watch that other issue (aimed for 2.0).

**Comment by fredfp** [comment]
@javanna what I cannot find is a setQuery() on SearchRequest. Same problem with ExistsRequest, I see no setQuery().

And yes, I use SearchRequestBuilder in multiple locations (via Client.prepareSearch()) then get the SearchRequest using SearchRequestBuilder.request() and all work fine.

**Comment by javanna** [comment]
@fredfp you need to do that in two steps:

```
SearchSourceBuilder searchSourceBuilder = SearchSourceBuilder.searchSource().query(/*your query goes here*/);
searchRequest.source(searchSourceBuilder);
```

or

```
QuerySourceBuilder querySourceBuilder = new QuerySourceBuilder();
querySourceBuilder.setQuery(/*your query goes here*/);
existsRequest.source(querySourceBuilder);
```

I hope this clarifies things. This confusion, as well as the issue that experienced when retrieving the request from builders will be cleared up with #9201. In the meantime I would recommend to use requests directly instead of builders in your case.

**Comment by fredfp** [comment]
Thank you @javanna!

I had never heard of SearchSourceBuilder but now this makes perfect sense to me and it indeed allows to completely avoid using request builders.

Now, I don't know if you want to keep this issue open. I still find it weird that using SearchRequestBuilder....setQuery().request() works but ExistsRequestBuilder...setQuery().request() doesn't.
It is not anymore an issue for me as you helped me work around it/use the API in a correct way. To me and as such, the current issue could also be closed/marked as duplicate of #9201.

Thanks!

**Comment by javanna** [comment]
thanks for your feedback @fredfp , I do agree that this is annoying, but since we might have the same problem in other apis and we are going to refactor all this with #9201 I will just close this issue in favour of #9201, where this problem will be taken care of.
